### ISSUE TITLE
E0004 says "note: `Option<i32>` defined here" when it's not

### ISSUE THREAD

**Issue by carols10cents** [rust-lang/rust#101068]
I just noticed this regression from 1.60 to 1.61 because TRPL reasons.

## Code

Given [the following code](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=24be8ca60112182b3300a9e7d4ce2e17): 

```rust
fn main() {
    let some_option_value: Option<i32> = None;
    let Some(x) = some_option_value;
}
```

The current output is:

```
error[E0005]: refutable pattern in local binding: `None` not covered
 --> src/main.rs:3:9
  |
3 |     let Some(x) = some_option_value;
  |         ^^^^^^^ pattern `None` not covered
  |
  = note: `let` bindings require an "irrefutable pattern", like a `struct` or an `enum` with only one variant
  = note: for more information, visit https://doc.rust-lang.org/book/ch18-02-refutability.html
note: `Option<i32>` defined here
  = note: the matched value is of type `Option<i32>`
help: you might want to use `if let` to ignore the variant that isn't matched
  |
3 |     let x = if let Some(x) = some_option_value { x } else { todo!() };
  |     ++++++++++                                 ++++++++++++++++++++++
```

Notice especially this line:

```
note: `Option<i32>` defined here
```

Which... raises a few questions... 

1. I definitely did not define `Option<i32>` in this code, so I don't know why this note is saying I did
2. Or is it saying that? The note doesn't point to what it thinks "here" is.
3. What is this note trying to tell me about, really?

Ideally the output shouldn't have this line, I don't think.

## Versions with regression

This problem started with 1.61 and still exists in 1.63.0, 1.64.0-beta.3, and nightly 2022-08-25 7480389611f9d04bd34a.

@rustbot modify labels: +regression-untriaged

**Comment by apiraino** [comment]
if case it's useful, the bisection points to the nightly-2022-03-09 and this rollup merge:
```
  commit[0] 2022-03-07UTC: Auto merge of #94706 - matthiaskrgr:rollup-l5erynr, r=matthiaskrgr
  commit[1] 2022-03-07UTC: Auto merge of #94709 - martingms:link-to-chunked-opt-pr, r=nnethercote
  commit[2] 2022-03-08UTC: Auto merge of #90887 - jackh726:issue-90729, r=nikomatsakis
  commit[3] 2022-03-08UTC: Auto merge of #94716 - RalfJung:miri, r=RalfJung
  commit[4] 2022-03-08UTC: Auto merge of #94734 - matthiaskrgr:rollup-28shqhy, r=matthiaskrgr
  commit[5] 2022-03-08UTC: Auto merge of #92260 - jyn514:less-python-logic, r=Mark-Simulacrum
  commit[6] 2022-03-08UTC: Auto merge of #94702 - b-naber:static-refs-mir, r=lcnr
```

two of the them are also rollups :roll_eyes: I think the most likely is 94734 (possibly #94580?)

**Comment by apiraino** [comment]
WG-prioritization assigning priority ([Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.23101068.20E0004.20says.20.22note.3A.20.60Option.3Ci32.3E.60.20defined.20here.22.20when.20.E2.80.A6)).

@rustbot label -I-prioritize +P-medium

**Comment by Jules-Bertholet** [comment]
Seems to have been fixed in 1.68: https://rust.godbolt.org/z/5no68WMPx

@rustbot label -regression-untriaged +regression-from-stable-to-stable

**Comment by apiraino** [comment]
then I'm going to close this as fixed (feel free to reopen if you disagree)
